config/v1/types_ingress.go

```diff
@@ -41,13 +44,103 @@ type IngressSpec struct {
+	//
+	// An operator that creates entries in this slice should clean them up during removal.
+	// An operator must also handle the case of deleted status without churn.
+	// +kubebuilder:validation:Optional
+	// +optional
+	ComponentRoutes []ComponentRouteStatus `json:"componentRoutes,omitempty"`
 }
 
-// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// ComponentRouteSpec allows for configuration of a route's hostname and serving certificate.
+type ComponentRouteSpec struct {
+	// Namespace is the namespace of the route to customize. It must be a real namespace. Using an actual namespace
+	// ensures that no two components will conflict and the same component can be installed multiple times.
+	// +kubebuilder:validation:Pattern=[a-z0-9]([-a-z0-9]*[a-z0-9])?
+	// +kubebuilder:validation:Required
+	// +required
+	Namespace string `json:"namespace"`
+
+	// Name is the logical name of the route to customize.
+	// +kubebuilder:validation:Required
+	// +required
+	Name string `json:"name"`
+
+	// Hostname is the hostname that should be supported by the route.
+	// +kubebuilder:validation:Required
+	// +required
+	Hostname string `json:"hostname"`
+
+	// ServingCertKeyPairSecret is a reference to a secret in namespace/openshift-config that is a kubernetes tls secret.
+	// The serving cert/key pair must match and will be used by the operator to fulfill the intent of serving with this name.
+	// +kubebuilder:validation:Required
+	// +required
+	ServingCertKeyPairSecret SecretNameReference `json:"servingCertKeyPairSecret"`
+}
+
+// ComponentRouteStatus contains information allowing configuration of a route's hostname and serving certificate.
+type ComponentRouteStatus struct {
+	// Namespace is the namespace of the route to customize. It must be a real namespace. Using an actual namespace
+	// ensures that no two components will conflict and the same component can be installed multiple times.
+	// +kubebuilder:validation:Pattern=[a-z0-9]([-a-z0-9]*[a-z0-9])?
+	// +kubebuilder:validation:Required
+	// +required
+	Namespace string `json:"namespace"`
 
+	// Name is the logical name of the route to customize. It does not have to be the actual name of a route resource.
+	// Keep in mind that this is your API for users to customize. You could later rename the route, but you cannot rename
+	// this name.
+	// +kubebuilder:validation:Required
+	// +required
+	Name string `json:"name"`
+
+	// DefaultHostname is the normal hostname of this route.
+	// +kubebuilder:validation:Required
+	// +required
+	DefaultHostname string `json:"defaultHostname"`
+
+	// ConsumingUsers is a slice of ServiceAccounts that need to have read permission on the ServingCertKeyPairSecret secret.
+	// +kubebuilder:validation:Required
+	// +required
+	ConsumingUsers []string `json:"consumingUsers"`
+
+	// CurrentHostnames is the current name used by the route.
+	// +kubebuilder:validation:Required
+	// +required
+	CurrentHostnames []string `json:"currentHostnames"`
```

Comment:
at least one?

Reply:
Done.

Reply:
`uniqueItems` cannot be used https://github.com/kubernetes/apiextensions-apiserver/blob/eaeda8805efb2a4373ea0cc7ae2f4771668e6c67/pkg/apis/apiextensions/validation/validation.go#L749

Reply:
> at least one?

If we don't have status back from a route yet, we don't know this value

Reply:
> If we don't have status back from a route yet, we don't know this value


@deads2k since each operator that opts into this feature writes the componentRoute to the status wouldn't the currentHostname always be the default hostname or the configured hostname?

Reply:
It would not be.  I can have a default of "foo.bar.com", but if the router hasn't accepted my route, I have no current name because I'm not available.

Reply:
Thanks for the knowledge share. Addressed in the latest commit.